tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -4618,6 +4627,136 @@ class HLSLExternalSource : public ExternalSemaSource {
+      return true;
+    case AR_OBJECT_RWTEXTURE2D_ARRAY:
+    case AR_OBJECT_ROVTEXTURE2D_ARRAY:
+      ResKind = DXIL::ResourceKind::Texture2DArray;
+      ResClass = DXIL::ResourceClass::UAV;
+      return true;
+    case AR_OBJECT_TEXTURE3D:
+      ResKind = DXIL::ResourceKind::Texture3D;
+      ResClass = DXIL::ResourceClass::SRV;
+      return true;
+    case AR_OBJECT_RWTEXTURE3D:
+    case AR_OBJECT_ROVTEXTURE3D:
+      ResKind = DXIL::ResourceKind::Texture3D;
+      ResClass = DXIL::ResourceClass::UAV;
+      return true;
+    case AR_OBJECT_TEXTURECUBE:
+      ResKind = DXIL::ResourceKind::TextureCube;
+      ResClass = DXIL::ResourceClass::SRV;
+      return true;
+    case AR_OBJECT_TEXTURECUBE_ARRAY:
+      ResKind = DXIL::ResourceKind::TextureCubeArray;
+      ResClass = DXIL::ResourceClass::SRV;
+      return true;
+    case AR_OBJECT_TEXTURE2DMS:
+      ResKind = DXIL::ResourceKind::Texture2DMS;
+      ResClass = DXIL::ResourceClass::SRV;
+      return true;
+    case AR_OBJECT_RWTEXTURE2DMS:
+      ResKind = DXIL::ResourceKind::Texture2DMS;
+      ResClass = DXIL::ResourceClass::UAV;
+      return true;
+    case AR_OBJECT_TEXTURE2DMS_ARRAY:
+      ResKind = DXIL::ResourceKind::Texture2DMSArray;
+      ResClass = DXIL::ResourceClass::SRV;
+      return true;
+    case AR_OBJECT_RWTEXTURE2DMS_ARRAY:
+      ResKind = DXIL::ResourceKind::Texture2DMSArray;
+      ResClass = DXIL::ResourceClass::UAV;
+      return true;
+    case AR_OBJECT_BUFFER:
+      ResKind = DXIL::ResourceKind::TypedBuffer;
+      ResClass = DXIL::ResourceClass::SRV;
+      return true;
+    case AR_OBJECT_RWBUFFER:
+    case AR_OBJECT_ROVBUFFER:
+      ResKind = DXIL::ResourceKind::TypedBuffer;
+      ResClass = DXIL::ResourceClass::UAV;
+      return true;
+    case AR_OBJECT_BYTEADDRESS_BUFFER:
+      ResKind = DXIL::ResourceKind::RawBuffer;
+      ResClass = DXIL::ResourceClass::SRV;
+      return true;
+    case AR_OBJECT_RWBYTEADDRESS_BUFFER:
+    case AR_OBJECT_ROVBYTEADDRESS_BUFFER:
+      ResKind = DXIL::ResourceKind::RawBuffer;
+      ResClass = DXIL::ResourceClass::UAV;
+      return true;
+    case AR_OBJECT_CONSUME_STRUCTURED_BUFFER:
+    case AR_OBJECT_APPEND_STRUCTURED_BUFFER:
+      // It may seem incorrect to make these SRV,
+      // but it is consistent with GetHLSLResourceProperties().
+    case AR_OBJECT_STRUCTURED_BUFFER:
+      ResKind = DXIL::ResourceKind::StructuredBuffer;
+      ResClass = DXIL::ResourceClass::SRV;
+      return true;
+    case AR_OBJECT_RWSTRUCTURED_BUFFER:
+    case AR_OBJECT_ROVSTRUCTURED_BUFFER:
+      ResKind = DXIL::ResourceKind::StructuredBuffer;
+      ResClass = DXIL::ResourceClass::UAV;
+      return true;
+    case AR_OBJECT_CONSTANT_BUFFER:
+      ResKind = DXIL::ResourceKind::CBuffer;
+      ResClass = DXIL::ResourceClass::CBuffer;
+      return true;
+    case AR_OBJECT_TEXTURE_BUFFER:
+      ResKind = DXIL::ResourceKind::TBuffer;
+      ResClass = DXIL::ResourceClass::CBuffer;
```

Comment:
Technically, `ResClass` for `TextureBuffer` should be `DXIL::ResourceClass::SRV`, not `CBuffer`.